Tools/Algorithms/psatd.ipynb

```diff
@@ -0,0 +1,631 @@
+    "# and related additional terms in the first-order time derivative of the electric field\n",
+    "if divE_cleaning:\n",
+    "    dEx_dt += I*c**2*F*kx \n",
+    "    dEy_dt += I*c**2*F*ky\n",
+    "    dEz_dt += I*c**2*F*kz\n",
+    "    dF_dt = I*(kx*Ex+ky*Ey+kz*Ez)\n",
+    "\n",
+    "# Define first-order time derivative of the scalar field G used with div(B) cleaning,\n",
+    "# and related additional terms in the first-order time derivative of the magnetic field\n",
+    "if divB_cleaning:\n",
+    "    dBx_dt += I*c**2*G*kx\n",
+    "    dBy_dt += I*c**2*G*ky\n",
+    "    dBz_dt += I*c**2*G*kz\n",
+    "    dG_dt = I*(kx*Bx+ky*By+kz*Bz)\n",
+    "\n",
+    "# Define array of first-order time derivatives of the electric and magnetic fields\n",
+    "dE_dt = Matrix([[dEx_dt], [dEy_dt], [dEz_dt]])\n",
+    "dB_dt = Matrix([[dBx_dt], [dBy_dt], [dBz_dt]])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Linear system of ODEs for $\\boldsymbol{E}$, $\\boldsymbol{B}$, $F$ and $G$:\n",
+    "$$\n",
+    "\\frac{\\partial}{\\partial t}\n",
+    "\\begin{bmatrix}\n",
+    "\\boldsymbol{E} \\\\\n",
+    "\\boldsymbol{B} \\\\\n",
+    "F \\\\\n",
+    "G\n",
+    "\\end{bmatrix}\n",
+    "= M\n",
+    "\\begin{bmatrix}\n",
+    "\\boldsymbol{E} \\\\\n",
+    "\\boldsymbol{B} \\\\\n",
+    "F \\\\\n",
+    "G\n",
+    "\\end{bmatrix}\n",
+    "-\\mu_0 c^2 \n",
+    "\\begin{bmatrix}\n",
+    "\\boldsymbol{J} \\\\\n",
+    "\\boldsymbol{0} \\\\\n",
+    "\\rho \\\\\n",
+    " 0\n",
+    "\\end{bmatrix}\n",
+    "$$"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# Define array of all fields\n",
+    "fields_list = [Ex, Ey, Ez, Bx, By, Bz]\n",
+    "if divE_cleaning:\n",
+    "    fields_list.append(F)\n",
+    "if divB_cleaning:\n",
+    "    fields_list.append(G)\n",
+    "EBFG = zeros(dim, 1)\n",
+    "for i in range(EBFG.shape[0]):\n",
+    "    EBFG[i] = fields_list[i]\n",
+    "\n",
+    "# Define array of first-order time derivatives of all fields\n",
+    "fields_list = [dEx_dt, dEy_dt, dEz_dt, dBx_dt, dBy_dt, dBz_dt]\n",
+    "if divE_cleaning:\n",
+    "    fields_list.append(dF_dt)\n",
+    "if divB_cleaning:\n",
+    "    fields_list.append(dG_dt)\n",
+    "dEBFG_dt = zeros(dim, 1)\n",
+    "for i in range(dEBFG_dt.shape[0]):\n",
+    "    dEBFG_dt[i] = fields_list[i]\n",
+    "dEBFG_dt = dEBFG_dt.expand()\n",
+    "\n",
+    "# Define matrix M representing the linear system of ODEs\n",
+    "M = zeros(dim)\n",
+    "for i in range(M.shape[0]):\n",
+    "    for j in range(M.shape[1]):\n",
+    "        M[i,j] = dEBFG_dt[i].coeff(EBFG[j], 1)\n",
+    "print(r'M = ')\n",
+    "display(M)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Diagonalization of $W_1 = M (t-t_n) = P_1 D_1 P_1^{-1}$:"
```

Comment:
Do we really need to calculate the diagonalization of $W_1$, $W_2$ and $W_3$?
It seems that diagonalizing $M$ only would be enough (and then you would multiply the resulting diagonal matrix by $(t-t_n)$, $t$ or $s$ to get the diagonalization of $W_1$, $W_2$ and $W_3$.

Reply:
Thank you for the suggestion. I just added in f20061d an automated verification of the solution of the linear system of ODEs. Now I'll be able to make changes like the ones you suggested while making sure that I do not introduce any bugs. Will try your suggestions and report on the results soon.

Reply:
Done in 86b93d4: we now diagonalize only $M$ as $P D P^{-1}$ and then use $P$ and $D$ for the diagonalization of $W_1$, $W_2$ and $W_3$.